### PR TITLE
Add Nav DB module

### PR DESCRIPTION
**Pull request by dt-exafore**
This change adds Nav DB module.

### CODE REVIEW

**Comment by rge-exafore** [comment]
swift-nav/libswiftnav#314 and piksi_firmware#629 are not needed for this PR.

File: src/ndb.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  almanac_t alma;
+  u32 l2c_capabilities;
+} ndb_element_t;
+
+/* Maximum waiting time for write request, milliseconds */
+#define NV_WRITE_REQ_TIMEOUT 100
+
+#define NDB_IE_DIRTY (1 << 0) /**< needs to be written to NVM */
+#define NDB_IE_VALID (1 << 1) /**< value in RAM has been set */
+#define NDB_IE_IS_VALID(md_ptr) (md_ptr->nv_data.state & NDB_IE_VALID)
```
Comment (dt-exafore) [inline src/ndb.h]: Not used.

File: src/ephemeris.c [hunk 2]
```diff
@@ -81,28 +33,17 @@ void ephemeris_new(ephemeris_t *e)
...
   u32 index = sid_to_global_index(e->sid);
-  if (!ephemeris_valid(&es[index], &t)) {
-    /* Our currently used ephemeris is bad, so we assume this is better. */
-    log_info("New untrusted ephemeris for %s", buf);
-    ephemeris_lock();
-    es[index] = es_candidate[index] = *e;
-    ephemeris_unlock();
 
-  } else if (ephemeris_equal(&es_candidate[index], e)) {
+  if(ephemeris_equal(&es_candidate[index], e)) {
```
Comment (dt-exafore) [inline src/ephemeris.c]: space

File: src/manage.c [hunk 3]
```diff
@@ -250,10 +246,11 @@ static u16 manage_warm_start(gnss_signal_t sid, const gps_time_t* t,
 
     /* Do we have a suitable ephemeris for this sat?  If so, use
        that in preference to the almanac. */
-    const ephemeris_t *e = ephemeris_get(sid);
-    if (ephemeris_valid(e, t)) {
+    ephemeris_t ephe;
+    enum ndb_op_code oc = ndb_ephemeris_read(sid, &ephe);
+    if((NDB_ERR_NONE == oc) && ephemeris_valid(&ephe, t)) {
```
Comment (dt-exafore) [inline src/manage.c]: space

File: src/manage.c [hunk 4]
```diff
@@ -532,9 +530,9 @@ static void manage_track()
     }
 
     /* Is ephemeris or alert flag marked unhealthy?*/
-    const ephemeris_t *e = ephemeris_get(sid);
-    /* TODO: check alert flag */
-    if (e->valid && !satellite_healthy(e)) {
+    u8 v, h;
+    enum ndb_op_code oc = ndb_ephemeris_info(sid, &v, &h, NULL, NULL);
+    if((NDB_ERR_NONE == oc) && v && !h) {
```
Comment (dt-exafore) [inline src/manage.c]: space

File: src/manage.c [hunk 5]
```diff
@@ -269,14 +266,15 @@ static u16 manage_warm_start(gnss_signal_t sid, const gps_time_t* t,
       if (time_quality >= TIME_FINE)
         dopp_uncertainty = DOPP_UNCERT_EPHEM;
     } else {
-      const almanac_t *a = &almanac[sid_to_global_index(sid)];
-      if (a->valid) {
-        calc_sat_az_el_almanac(a, t->tow, t->wn-1024,
+      almanac_t alma;
+      oc = ndb_almanac_read(sid, &alma);
+      if((NDB_ERR_NONE == oc) && (alma.valid)) {
```
Comment (dt-exafore) [inline src/manage.c]: space

File: src/base_obs.c [hunk 6]
```diff
@@ -300,9 +301,9 @@ static void obs_callback(u16 sender_id, u8 len, u8 msg[], void* context)
 
     /* Check if we have an ephemeris for this satellite, we will need this to
      * fill in satellite position etc. parameters. */
-    ephemeris_lock();
-    ephemeris_t *e = ephemeris_get(sid);
-    if (ephemeris_valid(e, &t)) {
+    ephemeris_t ephe;
+    if((ndb_ephemeris_read(sid, &ephe) == NDB_ERR_NONE) &&
```
Comment (dt-exafore) [inline src/base_obs.c]: space

File: src/decode_gps_l1.c [hunk 7]
```diff
@@ -112,10 +113,11 @@ static void decoder_gps_l1_process(const decoder_channel_info_t *channel_info,
...
   ephemeris_new(&e);
 
-  ephemeris_t *eph = ephemeris_get(channel_info->sid);
-  if (!eph->valid) {
+  u8 v, h;
+  ndb_ephemeris_info(channel_info->sid, &v, &h, NULL, NULL);
+  if (!v) {
     char buf[SID_STR_LEN_MAX];
     sid_to_string(buf, sizeof(buf), channel_info->sid);
-    log_info("%s ephemeris is invalid", buf);
```
Comment (jacobmcnamee) [inline src/decode_gps_l1.c]: rebase and use `log_info_sid()`

Comment (jacobmcnamee) [inline src/decode_gps_l1.c]: Actually I don't even think we should be checking this here. Maybe just call `ephemeris_new()` and delete the stuff after.

File: src/ephemeris.c [hunk 8]
```diff
@@ -81,28 +33,23 @@ void ephemeris_new(ephemeris_t *e)
...
   sid_to_string(buf, sizeof(buf), e->sid);
 
-  gps_time_t t = get_current_time();
+  if (!e->valid) {
+    log_error("Invalid ephemeris for %s", buf);
+    return;
+  }
+
   u32 index = sid_to_global_index(e->sid);
-  if (!ephemeris_valid(&es[index], &t)) {
```
Comment (jacobmcnamee) [inline src/ephemeris.c]: If the ephemeris in the NDB is invalid, we should store the incoming ephemeris to both the NDB and the candidate immediately. Only when the NDB already contains valid data should we wait for the candidate match before updating.

File: src/ephemeris.c [hunk 9]
```diff
@@ -121,42 +68,18 @@ static void ephemeris_msg_callback(u16 sender_id, u8 len, u8 msg[], void* contex
     log_warn("Ignoring ephemeris for invalid sat");
     return;
   }
-
-  ephemeris_new(&e);
```
Comment (jacobmcnamee) [inline src/ephemeris.c]: I think this should still call `ephemeris_new()`

File: src/solution.c [hunk 10]
```diff
@@ -417,20 +418,19 @@ static msg_t solution_thread(void *arg)
     static u8 n_ready_old = 0;
     u64 nav_tc = nap_timing_count();
     static navigation_measurement_t nav_meas[MAX_CHANNELS];
-
-    const channel_measurement_t *p_meas[n_ready];
-    navigation_measurement_t *p_nav_meas[n_ready];
-    const ephemeris_t *p_e_meas[n_ready];
+    ephemeris_t ephe_cache[MAX_CHANNELS];
```
Comment (jacobmcnamee) [inline src/solution.c]: I think `ephe_cache` should be static and the arrays of pointers should be on the stack.

File: src/solution.c [hunk 11]
```diff
@@ -417,20 +418,19 @@ static msg_t solution_thread(void *arg)
...
+    const channel_measurement_t *p_meas[MAX_CHANNELS];
+    static navigation_measurement_t *p_nav_meas[MAX_CHANNELS];
+    static ephemeris_t *p_e_meas[MAX_CHANNELS];
     for (u8 i=0; i<n_ready; i++) {
       p_meas[i] = &meas[i];
       p_nav_meas[i] = &nav_meas[i];
-      p_e_meas[i] = ephemeris_get(meas[i].sid);
+      ndb_ephemeris_read(meas[i].sid, &ephe_cache[i]);
```
Comment (jacobmcnamee) [inline src/solution.c]: Any way this returns an error? If so what happens?

File: src/solution.c [hunk 12]
```diff
@@ -417,20 +418,19 @@ static msg_t solution_thread(void *arg)
...
+      p_e_meas[i] = &ephe_cache[i];
     }
 
-    ephemeris_lock();
     calc_navigation_measurement(n_ready, p_meas, p_nav_meas,
-                                (double)((u32)nav_tc)/SAMPLE_FREQ, p_e_meas);
-    ephemeris_unlock();
+                                (double)((u32)nav_tc)/SAMPLE_FREQ, (const ephemeris_t**)p_e_meas);
```
Comment (jacobmcnamee) [inline src/solution.c]: Why is the cast necessary?

File: src/solution.c [hunk 13]
```diff
@@ -487,18 +487,19 @@ static msg_t solution_thread(void *arg)
           if (dgnss_soln_mode == SOLN_MODE_LOW_LATENCY &&
               base_obss.has_pos) {
 
-            ephemeris_lock();
-            const ephemeris_t *e_nav_meas_tdcp[n_ready_tdcp];
+            ephemeris_t ephe_cache[MAX_CHANNELS];
```
Comment (jacobmcnamee) [inline src/solution.c]: My vote is static.

**Comment by jacobmcnamee** [comment]
Added a few inline comments on the changes to existing files. I haven't gone through the NDB files themselves yet.

If we're having RAM issues on v2, #647 should save 4kB+

**Comment by rge-exafore** [comment]
Closing  this PR, right one is https://github.com/swift-nav/piksi_firmware/pull/621.

**Pull request closed without merging**